### ISSUE TITLE
How to run sshd as a windows service ?

### ISSUE THREAD

**Issue by imjakey** [Microsoft/WSL#612]
Steps to reproduce:

step 1:  
start bash

step 2:
start sshd server
sudo /etc/init.d/ssh start

step 3:
connected to sshd server from xshell (a powerful terminal emulator which better than windows cmd line).
if you don't have xshell , use putty or any other client instead

step 4:
closed the bash cmd window , xshell connection is also closed.

Question:
how do I  run sshd as a windows service ? 

thanks !

**Comment by fpqc** [comment]
I would suggest instead running urxvt in xming, because sshd needs to be run as root, and there are annoying problems that can arise with that.

Right now there is no support for running daemons as windows services because WSL is meant for developer workstations rather than servers, although some unprivileged daemons (like urxvtd) can be run silently with a little bit of workaround.

**Comment by therealkenc** [comment]
That is not even remotely accurate. There are no annoying problems with running sshd as root (`sudo service ssh start`); and even if there were, you can run sshd on an unprivileged port no different than urxvtd. For people uninterested in inviting all of X11 to the party, PuTTY and a dozen of other SSH clients are lighter weight and make more sense. I use the [Chrome Secure Shell Extension](https://chrome.google.com/webstore/detail/secure-shell/pnhechapfaindjhompbnflcldabbghjo?hl=en). Indeed, X clients have been explicitly identified as low priority for the WSL project, but do run, happily, as a side effect of coincidental socket and pty improvements.

**Comment by fpqc** [comment]
Yeah, but I run urxvtd on startup with run.exe, whereas to start the sshd service you have to type a p/w (because you are sudoing it "superuser do", i.e. running it as the root user).  I have a shortcut that does 

`run.exe bash.exe -c "urxvtd -q -o -f ; bash"`

 in the startup folder for my Windows Use (in addition to running Xming on startup), so I can launch urxvtc whenever I want, with the daemon running totally silently with
`bash.exe -c urxvtc`

**Comment by aseering** [comment]
Or you could modify your sudoers file to not require a password for that command.

**Comment by fpqc** [comment]
Yeah, that works too!  Good idea!

**Comment by aseering** [comment]
Make sure you have the syntax right?  Test it on an actual Ubuntu machine if you have one available; the sudoers file format can be finicky to get right.

(Disclaimer:  I haven't actually tried this particular command.  I've run other commands with passwordless sudo under WSL, though.

**Comment by fpqc** [comment]
@therealkenc @aseering, @Manouchehri says in the cmake thread that you also need to run sshd with Windows admin credentials as well in order to open up a socket for sshd. Have you guys found a way around that?

**Comment by Manouchehri** [comment]
Can a normal user can ever bind on a system port on Windows?

**Comment by fpqc** [comment]
@Manouchehri I think so.  Bittorrent clients can, and they can even make UPnP requests.

**Comment by Manouchehri** [comment]
BitTorrent (shouldn't) run on system ports (0-1024). Most clients bind on a UDP port in the dynamic port range.

**Comment by aseering** [comment]
@fpqc -- could you run on a higher port?  The problem is that the default sshd port is in the system-port range, as @Manouchehri said.  I often bind servers to port 2222, or some other higher-numbered port, to work around that limitation.

**Comment by therealkenc** [comment]
@fpqc, I'm not running WSL in terminal with Windows admin privileges. It binds to 22 fine with `sudo service ssh start` or however you feel like starting the daemon. You can also start it on 2222 like @aseering says, but there's a minor issue with perms on /dev/tty so you still have to run it as root.

**Comment by Manouchehri** [comment]
@therealkenc Permission issues on the Windows or Linux side? `/dev/tty` shows as owned by the user to me.

```bash
ubuntu@DESKTOP-3RQO5S5:~$ ls -l /dev/tty
crw------- 1 ubuntu tty 4, 1 Jul  2 22:23 /dev/tty
```

**Comment by fpqc** [comment]
@therealkenc It seems to work for me too doing that, but I have a problem connecting to it with PuTTY, and if I open another terminal and try to connect it appears to connect but then dies.
```
fpqc@FPQC-PC:~|⇒  ssh -vvv 127.0.0.1 
OpenSSH_6.6.1, OpenSSL 1.0.1f 6 Jan 2014
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 19: Applying options for *
debug2: ssh_connect: needpriv 0
debug1: Connecting to 127.0.0.1 [127.0.0.1] port 22.
debug1: Connection established.
debug3: Incorrect RSA1 identifier
debug3: Could not load "/home/fpqc/.ssh/id_rsa" as a RSA1 public key
debug1: identity file /home/fpqc/.ssh/id_rsa type 1
debug1: identity file /home/fpqc/.ssh/id_rsa-cert type -1
debug1: identity file /home/fpqc/.ssh/id_dsa type -1
debug1: identity file /home/fpqc/.ssh/id_dsa-cert type -1
debug1: identity file /home/fpqc/.ssh/id_ecdsa type -1
debug1: identity file /home/fpqc/.ssh/id_ecdsa-cert type -1
debug1: identity file /home/fpqc/.ssh/id_ed25519 type -1
debug1: identity file /home/fpqc/.ssh/id_ed25519-cert type -1
debug1: Enabling compatibility mode for protocol 2.0
debug1: Local version string SSH-2.0-OpenSSH_6.6.1p1 Ubuntu-2ubuntu2.7
debug1: Remote protocol version 2.0, remote software version OpenSSH_6.6.1p1 Ubuntu-2ubuntu2.7
debug1: match: OpenSSH_6.6.1p1 Ubuntu-2ubuntu2.7 pat OpenSSH_6.6.1* compat 0x04000000
debug2: fd 3 setting O_NONBLOCK
debug3: load_hostkeys: loading entries for host "127.0.0.1" from file "/home/fpqc/.ssh/known_hosts"
debug3: load_hostkeys: loaded 0 keys
debug1: SSH2_MSG_KEXINIT sent
Connection closed by 127.0.0.1
```

**Comment by Manouchehri** [comment]
Looks like we actually can bind to 22/TCP without Windows admin privs?

```bash
ubuntu@DESKTOP-3RQO5S5:~$ sudo strace netcat -l -p22 -w5
sudo: unable to resolve host DESKTOP-3RQO5S5
execve("/bin/netcat", ["netcat", "-l", "-p22", "-w5"], [/* 14 vars */]) = 0
brk(NULL)                               = 0x1441000
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
# ...snip...
brk(0x1462000)                          = 0x1462000
socket(PF_INET, SOCK_STREAM, IPPROTO_TCP) = 3
setsockopt(3, SOL_SOCKET, SO_REUSEADDR, [1], 4) = 0
bind(3, {sa_family=AF_INET, sin_port=htons(22), sin_addr=inet_addr("0.0.0.0")}, 16) = 0
listen(3, 1)                            = 0
accept(3, {sa_family=AF_INET, sin_port=htons(16627), sin_addr=inet_addr("127.0.0.1")}, [16]) = 4
close(3)                                = 0
poll([{fd=4, events=POLLIN}, {fd=0, events=POLLIN}], 2, 5000) = 1 ([{fd=4, revents=POLLIN}])
read(4, "Yep, this works.\n", 2048)     = 17
write(1, "Yep, this works.\n", 17Yep, this works.
)      = 17
poll([{fd=4, events=POLLIN}, {fd=0, events=POLLIN}], 2, 5000) = 0 (Timeout)
close(4)                                = 0
close(3)                                = -1 EBADF (Bad file descriptor)
close(3)                                = -1 EBADF (Bad file descriptor)
exit_group(0)                           = ?
+++ exited with 0 +++
```

**Comment by therealkenc** [comment]
@Manouchehri - I see what's going on with my /dev/tty now. I start bash.exe with user root and then `su ken` in root's .bashrc. WSL seems to set the /dev/tty owner to the bash.exe user. So my /dev/tty is root:tty not ubuntu:tty. But the real problem is perms 600 instead of 666. Anyway you can set phasers on ignore here. sshd on 2222 will work just fine as user ubuntu if you start bash.exe as ubuntu.

@fpqc - dunno what exactly is going on there. I just followed the secret handshake in #300 and never really looked back since then. In short:

```
ListenAddress 0.0.0.0
UsePrivilegeSeparation no
PasswordAuthentication yes
```

**Comment by fpqc** [comment]
@therealkenc Yeah that works, but any idea why adding my PuTTY key to known_hosts doesn't allow me to connect password-free from PuTTY?

**Comment by therealkenc** [comment]
Follow [these instructions](https://www.getfilecloud.com/blog/ssh-without-password-using-putty/) (or others like it) and you should be good.

**Comment by qris** [comment]
Add the following line to `/etc/sudoers` (using `sudo visudo` for safety):

    %sudo ALL=(ALL) NOPASSWD: /usr/sbin/sshd -D

And then use the Task Scheduler to create a new basic task with these settings:

* Run: _When the computer starts_,
* Action: _Start a program_,
* Program: `c:\Windows\system32\bash.exe`, 
* Arguments: `-c "sudo /usr/sbin/sshd -D"`

Test it by clicking on the _Task Scheduler LIbrary_, selecting your new task and clicking _Run_ on the right-hand side. A command prompt window should open and remain open.

Click on _Properties_ and under _Security_, change to _Run whether the user is logged on or not._

**Comment by fpqc** [comment]
@qris I would also suggest running it with run.exe to hide the console if necessary.

**Comment by imjakey** [comment]
Thank you everyone, you have helped me a lot.

I use two scripts to achieve my goals:

script 1 named startssh.bat, the contents is :
cd C:\Windows\System32
bash.exe -c "/usr/sbin/sshd -D"

script 2 named autostartssh.vbe, the contents is :
set ws=wscript.createobject("wscript.shell")
ws.run "startssh.bat /start",0

the script 2 can run script 1 like a deamon, then I add scipt 2 to a Windows Task Scheduler, it will run on the computer starts.
Now, I can connected to my sshd server from xshell, it works well !

**Comment by rodrymbo** [comment]
@qris - When I change the setting to "Run whether the user is logged in or not," the task no longer starts, (at least when I click on it and choose "Run"). It works fine when set to "Run only when user is logged on". Any thoughts? Can anyone else replicate this behavior in Task Scheduler, or make a task start properly with "whether user is logged on or not"?

**Comment by qris** [comment]
@fpqc @rodrymbo Sorry, I posted my comment too soon. My suggestion of running the process regardless of whether the user is logged in or not did not work, as you discovered, and I have not been able to make it work. I have edited my post to avoid suggesting this option.

**Comment by jaydenn** [comment]
Thanks for the help, but I'm having an odd problem.. When I try to ssh in from my remote ubuntu laptop, it does not recognize my credentials. However, somehow it recognizes my windows user credentials, and when I ssh in I am dropped into a cmd shell - not bash.. I'm very confused, any help would be appreciated.

**Comment by hacnet** [comment]
@rodrymbo  I've been able to **autostart sshd whether user is logged on or not** but only by stopping lxssmanager first, i.e. point taskschd to a batch script with the following content:

sc stop lxssmanager
c:\Windows\System32\bash.exe -c "/usr/sbin/sshd -D"

but (!) after my subsequent login, I cannot start bash unless I issue the same stop command.
It's feels really messy and I'm wondering why this is such a headache? sshd should be up by default It's almost quicker to install Cygwin.

**Comment by fpqc** [comment]
@hacnet I mean, it's software in development that has only been publicly available for 3 or 4 months (over which time it has advanced in leaps and bounds).  I think you have to accept that there will be rough edges for some time into the future.  

(For example, you may remember testing Mozilla Firebird back a million years ago, and that took quite a while to reach maturity (before falling off a cliff and murdering the interface (and version numbering system) in recent years), and that project wasn't a quarter as ambitious as this one.)

**Comment by hacnet** [comment]
@fpqc while I appreciate that dev environments may have some rough edges, you don't start off the project by breaking the fundamental access method to the brand new shell you're announcing.

**Comment by fpqc** [comment]
@hacnet The fundamental (currently supported) access method is local access through the bash.exe console wrapper (didn't you notice that the whole WSL session dies when you close the last one of these?). Since the console wrapper right now is pretty glitchy and problematic, your next choice is probably something like rxvt (or I think now xterm), which both run with minimal setup through X.  

SSH is quite a bit more complicated (it runs as a persistent daemon as root and as a linux "system service"), and offers very little functional improvement over an X terminal emulator.  The primary use case in mind with current builds is not really remoting in, and there are better or comparable options when you are on the local machine.

**Comment by benhillis** [comment]
@hacnet See my https://github.com/Microsoft/BashOnWindows/issues/834#issuecomment-238589089 for a little bit of detail on services.  As @fpqc points out this project is in its infancy.  It might be helpful for you to think of this GitHub and our User Voice as a way to play a part in the development process.  I'll be the first to admit there are many things that need improvement.

**Comment by hacnet** [comment]
@fpqc @benhillis Not sure why you switched to talking about X. As Windows already has windows, it's not something I missed. Bash however is. Since it should not be presumed that everyone has local access to their Windows 10 machines, I would say accessing the all new bash via sshd cannot be neglected or put in the "nice to have" box, because many efficient workarounds provide their own Linux emulation thus undermining the whole object of this new subsystem feature.

**Comment by fpqc** [comment]
@hacnet you can run an X terminal emulator with the Windows port of the Xorg-server project vcXsrv (or Colin Harrison's closed-ish source version Xming, which I hear supports some h/w acceleration stuff that vcXsrv does not yet support).  The main point of doing this is to circumvent the currently buggy-to-broken Windows console wrapper for the WSL process /bin/bash called bash.exe.  Installing WSL, rxvt, and vcXsrv is still much smaller and faster than installing a full cygwin environment or virtualization.

**Comment by aseering** [comment]
@fpqc -- if I understand correctly, @hacnet 's use case is accessing a remote Windows machine via ssh.  Presumably in order to get remote access to files and state on that machine?  (If you're just using WSL as a stateless remote Linux dialup, not to put down the WSL folks, but honestly a Linux machine is going to make a better Linux dialup than a Windows machine :-) )

Windows offers Remote Access as an alternate means of logging into a remote system.  But Remote Access is bandwidth-intensive and higher latency.  If what you need is a terminal, or you're just copying files around, IMO ssh (or sftp, with a graphical client) is a much more pleasant experience.

**Comment by hacnet** [comment]
@fpqc Surely I'm not the only one whose WLS incentive and/or expectation is to finally be free of multi-tiered installs. Or at least to reduce them. At work, an Ubuntu environment accessible via ssh could amalgamate a whole range of system administration that otherwise needs to be split up between Windows and Unix/Linux teams. I certainly agree with avoiding virtualization solutions that all but cripple local workstations, yet I've not noticed so many performance issues running Cygwin (on relatively modern hw), so replacing it with a suite of 3 doesn't seem appealing. Thanks for the suggestion though.

**Comment by hacnet** [comment]
@aseering At home I want the one server for all things. It cannot do without Windows because certain things simply don't get developed for Linux, while at work I would want server administration, historically focused on Unix, universally applied to include Windows nodes.

**Comment by fpqc** [comment]
@hacnet Sure, but you have to remember that since the initial release into the insider stream, the team has been under instruction from their management to focus mainly on the single-workstation/client single-user use case, not the server/multi-client one.  If you notice, WSL installs a completely separate system for each Windows 10 user, further stressing the single-user use case.  

However, as a consequence of the software doing the emulation where it does (in the kernel-space, as a (theoretically) co-equal subsystem to Win32 (both kernel subsystems see their syscalls translated into native NT kernel syscalls)), it turns out that it is possible to run a lot of software that was not included in the initially-targeted use case.  

Because the team (and their managers probably) have seen such an outpouring of enthusiasm for things like running sshd as a service (and other things like running the rxvt daemon in the background, xsettingsd, unity-settings-daemon, or whatever), they are planning to come up with an elegant way to do it, but it didn't make it into the anniversary update.  They chose to release the feature as a beta because there was a lot of interest in it, and they felt like there was enough work ready to go that devs might actually use it (but that they should be aware that it isn't finished yet by any means, which is why it has a big (beta) right after the name when you install it in programs & features).  

I guess all I'm saying is try to be a little patient and understand that the reason why your desired feature didn't make it into the anniversary update is that it simply wasn't planned at the start.  It seems like it may make it into some of this year's builds though!

**Comment by aseering** [comment]
It occurs to me that I haven't posted this link here yet:

http://wsl-forum.qztc.io/viewtopic.php?f=6&t=10

Forum post that's intended to summarize how to get sshd working under WSL.  Based on the discussion that initially started in this and other threads.

There seem to be a bunch of tickets being filed here about getting ssh working.  Would it make sense to consolidate them in the single forum thread, so that everyone sees the same information and those of us who help out with these tickets don't have to repeat things that have already been discovered?

**Comment by aseering** [comment]
@hacnet -- ah, that use case makes sense.  It's not WSL's initial stated use case, as @fpqc has described.  But I'm confident that the WSL folks are thinking hard about it.

**Comment by hacnet** [comment]
@aseering 
Just saw your reference to: [http://wsl-forum.qztc.io/viewtopic.php?f=6&t=10](http://wsl-forum.qztc.io/viewtopic.php?f=6&t=10) Another forum to sign up for?
`set ws=wscript.createobject("wscript.shell")` didn't work for me - Event Viewer came up with all kinds of dcom permission issues. The errors disappeared after painstakingly searching for, and taking ownership of, respective AppIDs then adjusting component execution rights. Yet my leaps of joy soon stopped when I saw the server still didn't start while no user was logged on. I'm also unlcear on the script's purpose (sorry no exposure to vbs), since sshd seemed to autostart without logged on users  without it - as long as a `sc stop lxssmanager` command was issued first.

**Comment by rodrymbo** [comment]
@hacnet  - It's not clear to me why creating a wscript.shell object would open Event Viewer. If Event Viewer was supposed to open, someone would have mentioned it, and perhaps remarked that the dcom permission issues are normal for Windows 10, as unlikely as that might seem, and most everyone has them, not just those running WSL. Perhaps one of the Microsoft folks here has a different perspective, but I think I recall someone saying to ignore the errors. Maybe at some point Microsoft plans to fix the problem, so our error logs aren't filled with ridiculous error messages, but that has nothing to do with WSL (that I've seen reported anyway).

As for bash.exe running when no one is logged in: there seems to be some confusion about whether that is actually workable. I saw the `sc stop lxssmanager` instruction, but haven't had an occasion to test it; for one reason, I thought I saw someone say that the lxssmanager service is configured for "Manual" start (on boot) and that it runs automatically when bash.exe is run, and shuts itself down soon after bash.exe exits. So why one would need to run the `sc stop lxssmanager` soon after boot but before anyone had invoked bash.exe, makes little sense. I'm not saying that it doesn't work for you, just that it is one more thing that doesn't make sense (to me). Plus it is a bit too much work for me right now to document what happens (or doesn't happen) before someone logs in. (Yes, I could do it. Maybe I'm just lazy.)

As for the vb script file to start bash.exe: the same script file could also run sc.exe first, if the wscript.shell object can be created on your machine and Event Viewer doesn't interrupt. If you want to experiment. But such experimentation amounts to hacking (in the good sense), looking into ways to make things work in ways they weren't supposed to work.

Anyway, github is a place to report "bugs". WSL was designed (for some reason) to make it hard to run scripts in the background, let alone when no one has yet logged in, so those are "features", not bugs. That means that [UserVoice](https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo) is where requests/suggestions/complaints about it need to be lodged. It's also why Adam started the forum. It is not official, just a place for people who want to make WSL do the kind of work we need done. Perhaps your input and suggestions there could help. For example, the "sc" hack has been mentioned as working sometimes, but causing problems other times. It'd be great to know its limitations, interactions, and perhaps a bit more about how it works or how to make it work.

It's clear that services, daemons, and background jobs are still problemmatic, but the Microsoft folks hereabouts say that it is on their list. I suppose that means another strategy is to set aside one's WSL experimentation for a couple of months, then come back and see if the "feature" has been changed so it does what you (we) want or need.

**Comment by hacnet** [comment]
@rodrymbo just to clarify... it was me (not the script) that opened Event Viewer, trying to find clues in Windows Logs as to why things didn't autostart. I thought I saw connection between the wscript and the dcom errors which is why I continued to look for solutions without it. Dcom errors might well have been red herrings whereby startup failures were caused not by ownership or launch privileges but because something that bash.exe needed was already running. This theory seemed to be confirmed when observing results when sc stop was included in the batch. I should however add that I've reinstalled lxss a few times. This might also cause varying results depending on how (un)clean it removes itself. I might retry the whole thing on a fresh OS when I have time.

**Comment by bunklung** [comment]
I just wanted to comment on the work @imjakey had done.
_"I use two scripts to achieve my goals:

script 1 named startssh.bat, the contents is :
cd C:\Windows\System32
bash.exe -c "/usr/sbin/sshd -D"

script 2 named autostartssh.vbe, the contents is :
set ws=wscript.createobject("wscript.shell")
ws.run "startssh.bat /start",0

the script 2 can run script 1 like a deamon, then I add scipt 2 to a Windows Task Scheduler, it will run on the computer starts."_

I had to edit my /etc/ssh/sshd_config to:
ListenAddress 0.0.0.0
UsePrivilegeSeparation no
PasswordAuthentication yes

Then I had to create a task and make certain it was run with elevated privileges otherwise ping does not work within bash or the SSH session.

Additionally, I had to change the bat file to include a different port (2222 for example):
bash.exe -c "sudo /usr/sbin/sshd -D -p 2222"

Why? Because for some reason SSH (CMD.exe *NOT* WSL/Bash) is enabled on Windows 10 which uses port 22. Perhaps this is because I use Power Shell too? I know this was planned, but I didn't expect for it to be turned on. Surprise!

Additionally I choose not to have the event start automatically. I have no trigger for the event, I just create a shortcut which calls the task when I what SSH/WSL to run in the background. Help found here: http://www.digitalcitizen.life/use-task-scheduler-launch-programs-without-uac-prompts

**Comment by gdh1995** [comment]
Since `/usr/sbin/sshd` needs the root privilege, this command may be helpful:
* `sudo chmod +s /usr/sbin/sshd`
* Then, I can run `/usr/sbin/sshd -D` directly from a user `gdh1995`.
* `nohup` also works well.

The `+s` mod means `sshd` will be always ran as root, no matter whatever user you are using.

**Comment by rodrymbo** [comment]
There's a script for that! `sudo /etc/init.d/ssh start`. Yes, you might want to add the NOPASSWD: option to the `sudo visudo` (sudoers) file, and yes it is a few extra characters to type. (Or put into a script file.) As an added bonus, you can do `sudo /etc/init.d/ssh stop` to gracefully stop the server, and you can issue the start command as often as you want because the script checks to make sure only one instance is running. One would think the people who worked on that script had thought of everything over the 20+ years it has been in development. Or one could invent a new way to start sshd.

Or instead of running sshd suid root, one could just use `lxrun.exe /setdefaultuser root` and run everything as root. That's how people used to do it in the old days, before they got bit by security problems or fumble-fingers, and, magically, a lot of linux people and distros started using unprivileged users and `sudo`. 

There's no reason not to experiment with suid on WSL, since no one else can use one's WSL instance (unless one shares one's computer with them). But there are good reasons to stay away from it on regular linux servers. 

One can use `sudo sshd -D` to good effect in some places where `sudo /etc/init.d/ssh start` doesn't get it exactly right. For example, in a script started by Task Scheduler, `bash.exe -c "sudo sshd -D"` (or something like that) will happily run almost forever mostly in the background, meaning you don't have to worry about accidentally closing the last bash.exe window and losing all background tasks that you hoped would finish. The same trick can work other ways too, such as a script like 

```
#! /usr/bin/bash
/etc/init.d/ssh start
/usr/sbin/cron -f
```

which launches sshd in the background, then cron in the foreground (cron's `-f` does essentially what sshd's `-D` does), so one has cron holding the Bash.exe window open forever. (There's no corresponding startup script for cron like `/etc/init.d/ssh`, though you could make one.) Store that in `/usr/local/sbin/` and run from sudo. Such a script could be a "temporary replacement `upstart` (or init)" until WSL figures out whether to support services and daemons and background tasks. (Well, okay, there may be an extra bash process kicking around, but it doesn't eat much and probably needs a home.)

**Comment by therealkenc** [comment]
Just a little heads up that @bunklung's solution opens up your whole filesystem to anyone with network access to your computer. It is running as root and listening on 0.0.0.0. It doesn't even matter if you do this for a your nonroot WSL user and listen on 2222, because you p0wn /mnt/c. 

Probably doesn't matter if you're on a home network behind your firewall and you trust your family. Just please understand that _no one In Real Life sets up sshd like this_. If you are offended by the login prompt, set up your public keys. Also consider using Dropbear because it is plain easier to set up and plays well with WSL using standard HOWTOs on the web.

**Comment by bunklung** [comment]
How is it any more insecure than having RDC on? SSH is still password protected as is PowerShell SSH instance.

**Comment by therealkenc** [comment]
The difference is RDC depends on windows security service. That sudoers setup means I can scrape your C drive if I have a smart phone in your house and you've given me your WiFi password.

**Comment by bunklung** [comment]
With what password and username? Authentication is on.

**Comment by therealkenc** [comment]
Maybe I am misunderstanding your set up. A you prompted when you ssh in or not? If not, then neither will anyone who you have given WiFi access to your network.
```
sudo visudo
username host = (root) NOPASSWD: /usr/sbin/sshd
```

**Comment by bunklung** [comment]
That sudo/visudo change is so that I can launch SSHD without a sudo password prompt. The script that launches SSHD is non-interactive.

**Comment by bunklung** [comment]
And yes, anyone who SSH's into my setup requires a username and password on both 22 (PowerShell) and 2222 (WSL).

**Comment by therealkenc** [comment]
Ah - I didn't pay enough attention and thought you were trying to end-run the login prompt like instructions elsewhere. Totally my bad. I'll edit my post upstream.

**Comment by bunklung** [comment]
NP, I am by no means a security expert, but I didn't want a big hole in my machine either.

**Comment by gdh1995** [comment]
In my tests yesterday on Win 10 build 14905.1000, I added a task with the `hide` option enabled, but there's still a command line window when it starts.

Although I've solved this problem by using `run.exe` metioned above, I still expect a better solution - I find that run.exe exits directly and the task's status will change from `runing` into `ready`.

The task I added is triggered whenever my account logs in.

**Comment by bunklung** [comment]
@gdh1995 I think that's the compromise you have to pick. If you launch it as VBE from task manager, there is no windows, but the task stays running indefinitely.

If you use the run command, you get the pesky window.

What I suggest is going a step further and doing what I did, but just create a task in the task manager, don't have it automatically launch. Instead create a windows shortcut which calls that specific task. Place that short cut in your startup fold. You should have no popup window, and task manager should say ready instead of running.

**Comment by rickywu** [comment]
I tried use bat and vbs, but after double click both of them just one windows flash past, and yes, can't connect though ssh

bat:
C:\Windows\System32\bash.exe -c "/usr/sbin/sshd -D"

vbs:
set ws=wscript.createobject("wscript.shell")
ws.run "C:\Windows\System32\bash.exe -c 'sudo /usr/sbin/sshd -D'",0 

where is the problem?

**Comment by rjstone** [comment]
Note that this is also described on Windows Developer Feedback as [Consider enabling cron jobs, daemons and background tasks](https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/13653522-consider-enabling-cron-jobs-daemons-and-backgroun)

Has anyone tried using a windows service wrapper like https://github.com/kohsuke/winsw ?

**Comment by fpqc** [comment]
@rjstone A couple of people tried starting bash with task manager on boot, but it caused problems for various reasons.

**Comment by rjstone** [comment]
I thought the problem with starting lxss processes was related to there being no logged-in user on startup. If that's the case then this could help, possibly:

![service control panel login](http://i.imgur.com/VU5zQ35.png)

But I don't understand windows stuff well enough to know if this makes any difference. Perhaps that "log in" still doesn't set up whatever environment lxss is wanting.

Currently I'm starting a daemon from the task scheduler on login and just dealing with the annoying console window it opens. Before trying anything else I thought I'd poke around and see if I could come up with anything better than the existing suggestions.

**Comment by rodrymbo** [comment]
Each Windows Account has its own WSL/Bash.exe environment. So while it might be possible to use Local System to start sshd, it would not be the User's home directory or other environment she'd see when she logged in ... if it is even possible to start it that way. The DrvFS (/mnt/c, etc) drives would be the same, but if the environment was running "as" Local System, file permissions on those NTFS drives would end up being unusual from the perspective of the Windows user.

There are a number of ways to avoid seeing or dealing with the Bash.exe windows that can open. The [Run Utility](http://www.straightrunning.com/projectrun.php) is one I've used. I've also used [VBscript](http://wsl-forum.qztc.io/viewtopic.php?f=6&t=10). In Task Scheduler, the "program/script" would be `d:\path\to\run.exe`, while the arguments might be something like `c:\Windows\System32\bash.exe  -c '~/bin/wsl-start-sshd.sh'`. Using a bash script adds a bit of flexibility; it's easy to add something else you want to run in the background and it's good practice. :P One might need to play with the path.

It's more convenient and not a huge security issue to change the line with sudoers to `%sudo   ALL=(ALL:ALL) NOPASSWD: ALL` with the addition of `NOPASSWD:`. That way a script you run from the command line doesn't ask for your password, so you can run it from the [Bash.exe command line](https://msdn.microsoft.com/en-us/commandline/wsl/reference?f=255&MSPPError=-2147217396). But in order to keep the Bash.exe window open (even though it is hidden, it is still there) you'd need to make sure the last thing in the script doesn't exit; that is, it can't be something that will become a daemon or run in the background, which you DO probably want all the other things in the script to do. So for example, the last line of the script could be `sudo cron -f` (which runs cron in the foreground), while the previous line could be `sudo /etc/init.d/ssh start` (which runs sshd as a daemon in the background, then immediately continues with the next line or closes Bash.exe if there is no next line).

Most of upstart is not running in WSL, at present. What is starting sshd when you type `service ssh start` is the `/etc/init.d/ssh ` script mentioned above, which is just a bash script.

**Comment by rjstone** [comment]
> Each Windows Account has its own WSL/Bash.exe environment. 

That's what I'd expect based on what I've read, but I was basically suggesting entering my own login credentials for the service to run as. Problems are:

1. Storing credentials seems less secure than waiting for a login. Doesn't seem worth it on a latop/desktop that's almost always logged in unless maybe it's just super important to get to a machine while logged out.

2. I don't know if this would work anyway because I don't know if it grabs the whole registry profile, etc, for the user and thus picks up the lxss environment or just sets the UID for determining permissions. The latter seems more likely.

I'm obviously not a windows expert.

I'm inclined to just use the same techniques you describe since it's not that important, but it would be interesting to know if a service wrapper would allow this to work with an existing user account even when not logged in just in case it's necessary.

**Comment by fpqc** [comment]
@rjstone Anyone concerned with security on Windows basically has to enable something very much like sudo in gp policy (always ask for password on a secure desktop).  This means popping up with a password window every time you need root.  

MS does not enforce this setting by default because people are lazy. So MS's default is the worst of both worlds, annoying but totally worthless for security (so much that UAC is not "considered a security boundary"

**Comment by rodrymbo** [comment]
When you showed the picture of Task Scheduler with the Local Service selected (rather than "This account"), I thought you were wondering if the Local Service approach might work, since no one had been able to get the "This Account" option to work. I guess I assumed too much. Sorry.

I think the developers have received a good number of requests for the WSL sshd to run before anyone logs in, and we might as well wait and see what approach they take to making it work. 

A service wrapper would have to invoke the whole WSL environment. That looks to me like pretty much the same thing as telling Task Scheduler to do it, since the only way (we know about) to get WSL going is to run Bash.exe. If someone has figured out how the Task Scheduler approach is deficient, I suppose a special purpose wrapper could be written to correct the deficiencies before starting Bash.exe ... if it is possible to do it at all.

The credentials for Task Scheduler tasks as well as services are pretty well protected. I'd assume access to the physical device is the determining factor. Anyway, if your own credentials entered into a Task Scheduler task are at risk, then, basically, so is everything else on your machine ...

**Comment by rjstone** [comment]
@fpqc Yes, it's more like a security hand rail than a security barrier.

@rodrymbo That's actually the UI for the Services snap-in. But unless windows services run in a dramatically different environment than task scheduler tasks then I suppose it wouldn't make any difference.

**Comment by fpqc** [comment]
@rjstone I dout running bash as local system will work properly

**Comment by benhillis** [comment]
@fpqc - I believe that is explicitly blocked.

**Comment by fpqc** [comment]
@benhillis =]

**Comment by kabinpokhrel** [comment]
What is the working solution then ?

**Comment by fpqc** [comment]
@kabinpokhrel Either use vbscript or run.exe to start bash.exe as a hidden window, and have it launch with a script that starts your daemons.

**Comment by rodrymbo** [comment]
There are [multiple solutions](http://wsl-forum.qztc.io/viewtopic.php?f=6&t=10) to having sshd (and other services/daemons) start when the user logs in. They stay running as long as you don't close the last bash.exe window (easiest if you use one of the tricks to hide it) or log out. Start them manually with one or more bash scripts (such as can be found in /etc/init.d) since upstart/systemd isn't running and the `service` command gives errors (but sort of works).

There seem to be no reasonable solutions (yet) to having sshd start before the user logs in (that is, just after the system boots, as if it were a real system service) or (probably) to allow a different user's WSL environment to run a script while the first user is still logged in. For that we need to wait until @benhillis and the other devs figure out (a) if they want to allow that and (b) what the best way is to provide the feature. They keep teasing us, but so far haven't committed one way or another.

**Comment by fpqc** [comment]
@rodrymbo There is a good solution imo for that.  Create a second unprivileged user, set up WSL, and then start WSL for that user as a task on system startup.  

When you want to use WSL, connect to the sshd with PuTTY.

**Comment by kabinpokhrel** [comment]
<main>
			<h1>SSH into "Bash on Ubuntu on Windows"</h1>
			<h2>Steps</h2>
			<ol>
				<li>
					<h3>First uninstall the <b>openssh-server</b></h3>
					<p>In order to SSH into the Windows Subsystem for Linux (WSL), we start off with uninstalling the <strong>openssh-server</strong>.</p>
					<code>sudo apt-get remove openssh-server</code>
				</li>
				<li>
					<h3>Reinstall the same program (openssh-server)</h3>
					<p><code>sudo apt-get install openssh-server</code></p>	
				</li>
				<li>
					<h3>Configure the SSH default installation to </h3>
					<p><code>sudo nano /etc/ssh/sshd_config</code> and disallow root login by setting <code>PermitRootLogin no</code></p>
				</li>
				<li>
					<h3>Then add a line beneath it that says:</h3>
					<p><code>AllowUsers yourusername</code></p>
					<p>and make sure <code>PasswordAuthentication</code> is set to <code>yes</code> if you want to login using a password.</p>
				</li>
				<li>
					<h3>Disable privilege separation by adding/modifying:</h3>
					<p><code>UsePrivilegeSeparation no</code></p>
				</li>
				<li>
					<h3>Restart the ssh service with <i>full restart</i> flag, like so</h3>
					<p><code>sudo service ssh --full-restart</code></p>
				</li>
				<li>
					<h3>Connect to your Linux subsystem from Windows using a ssh client like PuTTY.</h3>
				</li>
			</ol>
		</main>

**Comment by ylluminate** [comment]
Nice steps there @kabinpokhrel. Can we get some instructions on how to make the filesystem appear as a unix fs?  Eg, I'd like to be able to: `scp something me@hostname:/mnt/d/some_dir/`

**Comment by qris** [comment]
Note that Windows now [runs an SSH server](https://www.reddit.com/r/Windows10/comments/4w4sew/windows_10_has_a_native_ssh_server/) on port 22, so it may not be possible to get OpenSSH running on that port too. So you may have to change the `Port` in `/etc/ssh/sshd_config` to some other port number, such as `23` or `2222`.

You can tell if this is happening because you will need to log in using your Windows username and password (not your UNIX username and password) and when you do, you'll get a `cmd.exe` command prompt, not a Bash prompt.

**Comment by rodrymbo** [comment]
@ylluminate - Are you saying that the standard ssh server that is part of WSL doesn't do SCP for you?

I'm pretty sure I've used SCP with WSL's ssh server and it looked like the FS layout you describe.

Remember, the "other" ssh server is not the same as the ssh server within WSL. If you don't like how that other server behaves, you'd need to complain somewhere else. And, as @qris says, they can't both be running on the same port at the same time. But then no ssh server should listen on port 22 anyway.

**Comment by gdh1995** [comment]
On the newest WSL, `-D` can be removed and `sudo /usr/sbin/sshd` is just OKay.

**Comment by fpqc** [comment]
@gdh1995 Even if you drop your last bash.exe Window?  That has always been the problem.

**Comment by gdh1995** [comment]
@fpqc you are right.

Oh my god, the single `sshd` process exits at once if I exit bash.exe.

**Comment by fpqc** [comment]
@gdh1995 Yeah, that was always the problem. I didn't even know about the sshd -D thing.

**Comment by Luindil** [comment]
When I follow kabinpokhrel's steps, I am not able to login into my computer it always give me 
"Permission denied, please try again." Even though I am 100% sure that the passowrd is correct, I created new Users and it still did not work. What is the reason for this?

**Comment by rodrymbo** [comment]
@Luindil - sounds like something odd is going on. You can try `ssh -v -p [portnumber] [username]@localhost` from a different bash.exe window to see more info, maybe it will tell you at what point it is failing. (Spelling everything out on the command line helps check assumptions, and -v adds information that might be helpful.)

To confirm the password, you could try `su - [username]` which will ask for the password, unless you are root. Or, it should be asking for the password when you use sudo (until you added the NOPASSWD: setting with visudo). (You do know it is not necessarily the one you use to log in to Windows, right?) To reset the password, you can use `lxrun /setdefaultuser root` before logging in, then the `passwd` command, then /setdefaultuser back again.

If the permission denied error is from something else, e.g. a file permission, knowing which file, might make solving the issue easier, though if you used `sudo` to start it that shouldn't be a problem.

If you don't get enough information with ssh -v, or from up to three of the -v's, you might need to look into ways to get sshd to put its log somewhere useful, like see if it puts its log to stdout if you run sshd in the foreground. I've still not gotten syslog (or systemd's journal) working...

The `service` command isn't really supported completely (currently), so you could bypass that and just run the bash script `/etc/init.d/ssh start|stop|restart`. `Service` usually works to start that bash script, but by going directly to the script you bypass one possible source of errors. You could also make a copy of the script and add command line options (e.g. logging?) to troubleshoot.

**Comment by lypanov** [comment]
@Luindil Make sure you disable the SSH related Windows Services that appear to be installed by default (namely, SSH Broker and SSH Proxy).

**Comment by Luindil** [comment]
Thank you both for answering, I just figured out that it is because the port is already used by windows. When I use e.g. 23 as a port it works like a charm. @qris already mentioned it in an above post.

**Comment by fpqc** [comment]
@luindil isn't port 23 something super important like smtp?

**Comment by sysworx** [comment]
This methods are not working for starting a apache2 for example...cause it starts and get closed after start...?

**Comment by aseering** [comment]
@sysworx -- there's more discussion of this approach here:  https://wsl-forum.qztc.io/viewtopic.php?f=6&t=10 .  Read through the comments; in particular, if you have a command that exits immediately, you have to somehow artificially add another command that doesn't exit immediately.

**Comment by PalashBansal96** [comment]
I also had to enable the ssh port in Windows Firewall for all incoming connections to allow remote logins.

**Comment by rickywu** [comment]
Hi all, 
I found this tool is very useful and can run ssh server without any window in background

**Comment by fenchu** [comment]
Gentlemen, We have for years been running sshd under cygwin. We use a cygrunserv process that runs as local administrator and map all registry settings. Cant we just get a local admin that starts the cron scheduler/services.

**Comment by fpqc** [comment]
@fenchu not yet. since distros are installed at the user rather than computer level, running as local admin will work weird. Running as your own user but with local admin permission could work but comes with its own problems.

**Comment by therealkenc** [comment]
`sshd` and other processes not attached to a tty now persist in WSL as of April Update.
